lib/http3/frame.c

```diff
@@ -59,3 +59,37 @@ int h2o_http3_decode_priority_update_frame(h2o_http3_priority_update_frame_t *fr
 
     return 0;
 }
+
+size_t h2o_http3_goaway_frame_capacity(quicly_stream_id_t stream_or_push_id)
+{
+    return 1   /* type */
+           + 1 /* length field. length should be less than 64, so 1 byte should be enough to represent it */
+           + quicly_encodev_capacity(stream_or_push_id);
+}
+
+uint8_t *h2o_http3_encode_goaway_frame(uint8_t *dst, quicly_stream_id_t stream_or_push_id)
+{
+    *dst++ = H2O_HTTP3_FRAME_TYPE_GOAWAY;                /* type */
+    *dst++ = quicly_encodev_capacity(stream_or_push_id); /* payload length */
+    dst = quicly_encodev(dst, stream_or_push_id);
+
+    return dst;
+}
+
+int h2o_http3_decode_goaway_frame(h2o_http3_goaway_frame_t *frame, const uint8_t *payload, size_t len, const char **err_desc)
+{
+    const uint8_t *src = payload, *end = src + len;
+
+    /* quicly_decodev below will reject len == 0 case */
+    if (len > PTLS_ENCODE_QUICINT_CAPACITY)
+        goto Fail;
```

Comment:
I think this error check is insufficient. Consider the case where the given payload was `0100` (len = 2 bytes). This should be considered invalid, as the second byte is a junk.

I think the correct thing to do in this function is to assert `src == end` after returning successfully from `quicly_decodev`.

Reply:
👍 I totally missed that case. Thanks!